File: backend/rtofolio/app/schemas/portfolio.py
from typing import Any, Dict, List, Optional, Union
from pydantic import BaseModel

class Theme(BaseModel):
    style: Optional[str] = None
    color_palette: Optional[str] = None
    font: Optional[str] = None
    tone: Optional[str] = None
    # Updated keys to snake_case
    max_width: Optional[str] = "1280px" 
    width: Optional[str] = "100%"
    margin: Optional[str] = "0 auto"
    display: Optional[str] = "flex"
    flex_direction: Optional[str] = "column"
    align_items: Optional[str] = "center"

class Slot(BaseModel):
    name: str
    type: Optional[str] = None
    optional: Optional[bool] = False

class Layout(BaseModel):
    type: Optional[str] = None
    direction: Optional[str] = None
    gap: Optional[str] = None
    align: Optional[str] = None
    justify: Optional[str] = None
    
    # Added sizing/spacing keys found in your JSON
    width: Optional[str] = None
    max_width: Optional[str] = None
    margin: Optional[str] = None
    
    # Supports {"desktop": 5, "mobile": 1}
    columns: Optional[Dict[str, int]] = None
    
    # Supports nested slots for Nav/Footer
    slots: Optional[Dict[str, Union[List[Slot], List[str], str]]] = None
    
    # Allow for nested layouts (desktop/mobile) if strict typing is needed, 
    # though usually handled at the Section level for polymorphism.
    desktop: Optional['Layout'] = None
    mobile: Optional['Layout'] = None

class ItemLayout(BaseModel):
    type: str
    layout: Optional[Layout] = None
    slots: Optional[List[Slot]] = None
    constraints: Optional[Dict[str, Any]] = None

class Section(BaseModel):
    type: str
    priority: Optional[str] = None
    
    # Can be a direct Layout object OR a dict with "desktop"/"mobile" keys
    layout: Union[Layout, Dict[str, Layout]]
    
    props: Optional[Dict[str, Any]] = None
    items: Optional[List[Dict[str, Any]]] = None
    
    # Updated key to snake_case
    item_layout: Optional[ItemLayout] = None

class PortfolioSchema(BaseModel):
    # Updated key to snake_case
    schema_version: str
    theme: Theme
    sections: List[Section]

# Necessary for recursive models (if Layout references itself)
Layout.update_forward_refs()

File: frontend/jobfolio-frontend/types/portfolio.ts
// Columns are numeric counts in the JSON (e.g. desktop: 1)
export type Columns = {
    desktop: number;
    mobile: number;
};

// A slot may be defined as a detailed object or just a string identifier
export type Slot = {
    name: string;
    optional?: boolean;
    type?: string;
};

// Slots map used in responsive desktop/mobile layout variants
export type SlotsMap = Record<string, (Slot | string)[] | string | null>;

// Responsive config used under `desktop` / `mobile` keys in some layouts
export type ResponsiveConfig = {
    type: string;
    gap?: string | null;
    align?: string | null;
    justify?: string | null;
    slots?: SlotsMap | null;
    // allow additional arbitrary keys (flexible schema coming from backend)
    [key: string]: any;
};

export type Theme = {
    width?: string;
    maxWidth?: string;
    margin?: string;
    display?: string;
    flexDirection?: string | null;
    alignItems?: string;
    colorPalette?: string;
    font?: string | null;
}


export type PortfolioSchema = {

    schemaVersion? : string | null;
    theme : Theme;
    sections : [];
};

// General layout that covers both simple and responsive shapes
export type Layout = {
    type?: string;
    direction?: string | null;
    gap?: string | null;
    columns?: Columns | null;
    desktop?: ResponsiveConfig | null;
    mobile?: ResponsiveConfig | null;
    width?: string | null;
    maxWidth?: string | null;
    margin?: string | null
    [key: string]: any;
};

export type Constraints = {
    maxWidth?: string | null;
};

export type ItemLayout = {
    type?: string;
    layout?: Layout | null;
    slots?: Slot[] | null;
    constraints?: Constraints | null;
};

// Project item shape based on provided JSON
export type ProjectItem = {
    id?: string;
    title?: string;
    description?: string;
    image?: string | null;
    linkButton?: string | null; // JSON uses `linkButton`
    // accept legacy or alternative field names
    link?: string | null;
    [key: string]: any;
};

export type Projects = {
    priority?: string | null;
    layout?: Layout | null;
    itemLayout?: ItemLayout | null;
    items?: ProjectItem[] | null;
};

// Specific props for the `hero` section (note: key with dash is allowed)
export type HeroProps = {
    name?: string | null;
    "hero-text"?: string | null;
    CTA?: string | null;
    image?: string | null;
    [key: string]: any;
};

export type Hero = {
    type: "hero";
    priority?: string | null;
    layout?: Layout | null;
    props?: HeroProps | null;
    items?: null;
    itemLayout?: null;
};

